### DISCUSSION TITLE
Added: LDM Notes for May 17th, 2021

### DISCUSSION THREAD

**Discussion by 333fred** [dotnet/csharplang#4760] [LDM Notes]
https://github.com/dotnet/csharplang/blob/main/meetings/2021/LDM-2021-05-17.md

## Agenda

1. Raw string literals

**Comment by CyrusNajmabadi** [comment]
> I'm worried about this. When I use VS Code with the extension to trim trailing whitespace on save, it doesn't understand string literals IIRC. It would be nice to have at least the option of an entirely empty line as an acceptable alternative to the full whitespace prefix.

That's the logic we'll be having.

**Comment by MgSam** [comment]
> We considered whether to use ``` or ''' instead of or in addition to """...

This seems like the wrong thing to consider. Given backticks have no current use in C#, it seems like ` vs """ is the right question to consider. 

> The ` symbol is concerning because it can be hard for non-English keyboard layouts to hit

I still don't buy this argument. I hope the developers we're trying to save from using backticks in C# never have to write JavaScript, Swift, Markdown, commands on Unix, F#, Powershell, or [any of the other languages that use them](https://en.wikipedia.org/wiki/Grave_accent#Use_in_programming). In other words, if I'm a developer using a non-English keyboard, I already need a shortcut to create backticks regardless of any decision the C# LDT makes. This character is far too commonly used to go out of mainstream usage now.

Having a single line regex that requires **6** double quotes (12 glyphs!) just to not deal with escaping is going to be really ugly. `"""^\S+@\S+\.\S+$"""` 🤮 

Really, the whole need for this feature is because verbatim strings in C# were misdesigned from the get-go and just traded one kind of escaping for another. If we must add yet another new string form, it should be as lightweight and easy to read as possible.

**Reply by 333fred** [reply]
There is some merit to the argument that we should use just a single `` ` ``. @CyrusNajmabadi, we didn't bring it up at LDM, did you consider it during the proposal process and rule it out?

**Reply by 333fred** [reply]
FWIW, though, I don't find the 6 double quotes (which is still 6 glyphs) vs 2 markdown ticks to be a particularly convincing argument. Your example looks fine to me.

**Reply by jnm2** [reply]
There would be some nice symmetry with markdown:
````
var testSyntax =
    ```cs
    Console.WriteLine("``");
    ```;

var regex = new Regex(`\d["']`);
````

**Reply by CyrusNajmabadi** [reply]
> did you consider it during the proposal process and rule it out?

Yes.  The issues were:

1. it didn't save enough to be valuable.
2. it made it appear as if we supported markdown, which we don't.
3. it is inconsistent with all other strings.  
4. it is unpleasant on some keyboards.

So, it effectively saves a couple of characters.  But that's really just not that valuable to warrant an entirely new quoting style.

**Reply by CyrusNajmabadi** [reply]
> Having a single line regex that requires 6 double quotes (12 glyphs!) just to not deal with escaping is going to be really ugly.

You don't need that.  You can represent that already with: `@"^\S+@\S+\.\S+$"`.  You only need raw strings if you don't want to have to escape `""` itself.  And, in that case, i actually think the regex reads completely fine.  

> Really, the whole need for this feature is because verbatim strings in C# were misdesigned from the get-go and just traded one kind of escaping for another.

I don't see how that changes anything in the discussion.  We're here.  Verbatim strings are 20 years old and aren't going anywhere.  We're not litigating the past, we're deciding on what a nice future is going forward.

**Reply by CyrusNajmabadi** [reply]
> There would be some nice symmetry with markdown:

Are core issue that came up is that we do not want to somehow imply that we are supporting markdown.  Or, put another way, we literally *may* want to support markdown in the future, in which case we'd likely want to use ` ```  ` and have that both be distinct from `"""` strings, but also follow as much sensible rules of the markdown spec as possible.  Having two forms of `` ` `` strings in the language (one which is 'markdown', and one which is 'raw') would be very unfortunate.

**Reply by qrli** [reply]
> There would be some nice symmetry with markdown:
> var testSyntax =
    ```cs
    Console.WriteLine("``");
    ```;

But we often need to put C# code into markdown ``` block right? How would markdown work if we put that code into markdown?  Like above which is incorrectly rendered.

**Reply by ghord** [reply]
I'm in agreement what we should consider single ` as a quotation character. Addressing some points:

> 1. it didn't save enough to be valuable.

The feature we are talking about is 100% about readability. And single line ```Write(`Hello world`)``` seems much more readable than `Write("""Hello world""")`. It also is much less likely to appear in any quoted raw string literals making it more clear which quotes are part of the string and which are c# quotes.

It also makes it easier to read in context of language fences (if they are ever considered)
```
var query = database.Query(sql `select * from User where Name = "Joe"`);
```
Seems very readable to me.

> 2. it made it appear as if we supported markdown, which we don't.

Single backtick would not be confused with markdown anymore than shell backtick quoted literals commands are. Also, let's face it there is no possible way that c# will support inline markdown any more than there is going to be support for inline xml outside of comments.
> 
> 3. it is inconsistent with all other strings.

I think triple quote would also be inconsistent with other string, as it is entirely new concept.
> 4. it is unpleasant on some keyboards.

Many languages include those already, programmers have been dealing with that for as long as there are shell scripts.

**Reply by alrz** [reply]
If we're talking about keyboards now, I also think a singe ` is a lot easier to type than """ over and over again. It looks like we're escaping a single quote but we are not. As others mentioned, a few other languages are already using backticks, which means you already have the muscle memory and a keyboard.

**Reply by HaloFour** [reply]
A few other languages also use `"""` for text blocks, such as Java, Kotlin, Groovy, Python and Swift.  I've never found it onerous to type or interpret.  As mentioned the syntax and behavior here was influenced by Java feature.

I think the argument for `` ` `` for single line raw strings might be more compelling since the contents are likely to be fairly short and the delimiter could take up a significant proportion of the expression.  To me it seems like that would be the choice when you'd want verbatim strings but want to avoid doubling `"`s to escape them.

**Reply by MgSam** [reply]
> FWIW, though, I don't find the 6 double quotes (which is still 6 glyphs)...

I was trying to find the correct technical name for the fact that a double quote is composed of two independent lines but couldn't come up with one :)

**Reply by MgSam** [reply]
> You don't need that. You can represent that already with: `@"^\S+@\S+\.\S+$"`. You only need raw strings if you don't want to have to escape `""` itself. And, in that case, i actually think the regex reads completely fine.

You're being super-nitpicky here. Obviously this specific example doesn't _require_ the new string form for readability, but many others will. Here's a simple regex that could be used for pulling the content out of a quoted string (using regex groups) `"""""(.*?)"""""`. Compare that to \`"(.*?)"\` Clearly the second is far more readable.
 
> > Really, the whole need for this feature is because verbatim strings in C# were misdesigned from the get-go and just traded one kind of escaping for another.
> 
> I don't see how that changes anything in the discussion. We're here. Verbatim strings are 20 years old and aren't going anywhere. We're not litigating the past, we're deciding on what a nice future is going forward.

You're missing the point. If we add a new string form it should be as widely applicable as possible. It should make verbatim strings obsolete. A lot of the feedback you've received (interpolation, single-line use). Is to use this new form in as many places as possible. If this feature is designed correctly there is no scenario I can think of where you'd want to ever use verbatim strings as they are inferior in every respect. Making the new strings uglier to use than verbatim strings muddies this story quite a bit.

You guys keep talking about making "learner" scenarios easier. Well you absolutely do not want to have to teach learners there are 3 different string forms each of which have different use cases. Relegating verbatim strings to an appendix at the back of the book alongside `delegate` for obsolete features sounds a lot better IMO.

**Reply by HaloFour** [reply]
`"""""(.*?)"""""` would be interpreted as `(.*?)`, not `"(.*?)"`.  The five double-quotes would be interpreted to be the starting/ending delimiter, and there is no mechanism to escape content within the string.

**Reply by orthoxerox** [reply]
@HaloFour What's the solution to encoding `"(.*?)"`? A multiline raw string?

**Reply by HaloFour** [reply]
Correct, the following three would be identical:

```
var pattern1 = "\"(.*?)\"";
var pattern2 = @"""(.*?)""";
var pattern3 = """
               "(.*?)"
               """;
```

**Reply by jnm2** [reply]
@orthoxerox Yes, the meeting notes call this out. A raw string with a value starting with `"` must be mulitiline.

**Reply by CyrusNajmabadi** [reply]
> seems much more readable than Write("""Hello world""")

You wouldn't write that. You'd write `Write("Hello World")`

**Reply by CyrusNajmabadi** [reply]
> Single backtick would not be confused with markdown

The concern is that we *do* potentially want to support markdown in the future. And now that would be confusing if we used backtick in multiple contexts with different rules.

**Reply by jnm2** [reply]
> > There would be some nice symmetry with markdown:
> > var testSyntax =
> > `` cs     Console.WriteLine("``");      ``;
> 
> But we often need to put C# code into markdown ``` block right? How would markdown work if we put that code into markdown? Like above which is incorrectly rendered.

@qrli The fact that the above is incorrectly rendered doesn't reflect on what I was saying. It was improperly copied/quoted. To answer your question, this is how you escape code blocks that contain ```` ``` ```` in markdown:

`````markdown
````markdown
```cs
Console.WriteLine("``");
```
````
`````

And this is how you escape inline code in markdown:
````markdown
`` `Some text` ``
````

**Reply by CyrusNajmabadi** [reply]
> Also, let's face it there is no possible way that c# will support inline markdown

I literally do not see why that would be the case. Especially as it's something that several ldm members (including myself) are interested in. I'd far prefer to have markdown instead of xml doc comments.

**Reply by jnm2** [reply]
@CyrusNajmabadi I don't see how that overlaps with inline string literal syntax though.

**Reply by CyrusNajmabadi** [reply]
Note that looking like markdown would itself be very misleading. People would assume they could write ``` ` `` ` ```. But that's not how this works.

**Reply by jnm2** [reply]
@CyrusNajmabadi I don't understand your point about ``` ` `` ` ```. People wouldn't expect that coming from markdown because markdown requires ```` ``` `` ``` ````.

**Reply by CyrusNajmabadi** [reply]
> Compare that to `"(.*?)"` Clearly the second is far more readable.

This is called out in the drawbacks section, with a suggestion on how we could support it. For example: `'"(.*?)"'`. 

Note that your approach makes some strings nicer, but would similarly make some harder because you still have the problem of: what if my content starts or ends with my delimiter?

So, for example, a regex matching a markdown fence would be more complex in your case as well and either require multi line. 

That said, as the proposal makes clear, this is something that can be expanded on to make the case of "I want to be single line and I want to start and end in a delimiter" easier in the future.

**Reply by CyrusNajmabadi** [reply]
> People wouldn't expect that coming from markdown because markdown requires 

It doesn't. You can write ``` ` `` ` ``` today. 

If you do, you get ` `` `.   Funky hunh!

**Reply by jnm2** [reply]
@CyrusNajmabadi TIL, thanks! https://spec.commonmark.org/0.29/#example-330

**Reply by CyrusNajmabadi** [reply]
> You're missing the point. If we add a new string form it should be as widely applicable as possible

This is goals 1 and 2. This proposal meets those goals. 

Note that replacing quote with backtick just replace one set of bad cases with a good set, but also does the reverse. That's why the spec states that in order to make things nice for the union of those cases, you'd need to allow for at least two delimiter forms. And even then, that would only help when the string started or ended with the other delimiter.
